File: examples/agent_loop.ts
/**
 * remem Agent Loop Example (TypeScript)
 * 
 * This script demonstrates the typical memory lifecycle of an AI agent:
 * 1. Store: The agent learns a new fact during a conversation.
 * 2. Recall: The agent retrieves relevant facts to answer a question.
 * 3. Consolidate: The agent processes its working memory into long-term graphs and resolves contradictions.
 * 
 * Usage:
 * Ensure the remem API is running locally: `cargo run -p remem-api`
 * Then run this script: `npx tsx examples/agent_loop.ts`
 */

import { Memory } from "../sdk/typescript/src/index.js";

async function main() {
  console.log("=== remem TypeScript Agent Loop Simulation ===\n");

  // Initialize the memory client
  // By default, it connects to http://localhost:7474 and reads REMEM_API_KEY if present
  const memory = new Memory({ project: "ts-agent-demo" });

  const sessionId = "session_ts_01";

  // --- Step 1: Store ---
  console.log("1. Storing new observations...");
  
  const obs1 = "The user prefers to write backend code in Rust.";
  console.log(`   -> Storing: "${obs1}"`);
  const r1 = await memory.store(obs1, { tags: ["user_preference", "coding"] });
  console.log(`      Stored with ID: ${r1.id} (Importance: ${r1.importance})`);

  const obs2 = "The user is currently evaluating Google Gemini 1.5 Flash for reasoning.";
  console.log(`   -> Storing: "${obs2}"`);
  const r2 = await memory.store(obs2, { tags: ["current_task", "llm"] });
  console.log(`      Stored with ID: ${r2.id} (Importance: ${r2.importance})\n`);

  // --- Step 2: Recall ---
  console.log("2. Recalling relevant context...");
  const query = "What languages does the user like for backend development?";
  console.log(`   -> Query: "${query}"`);
  
  const results = await memory.recall(query, { limit: 2 });
  console.log(`   -> Found ${results.length} relevant memories:`);
  for (const res of results) {
    console.log(`      - [Similarity: ${res.similarity.toFixed(2)}] ${res.content}`);
  }
  console.log();

  // --- Step 3: Consolidate ---
  console.log("3. Consolidating session memory...");
  console.log(`   -> Triggering consolidation for session: ${sessionId}`);
  
  try {
    const report = await memory.consolidate(sessionId, "gemini-1.5-flash");
    console.log("   -> Consolidation Complete:");
    console.log(`      New facts extracted: ${report.new_facts}`);
    console.log(`      Updated facts: ${report.updated_facts}`);
    console.log(`      Contradictions resolved: ${report.contradictions.length}`);
    console.log(`      Knowledge graph updates: ${report.knowledge_graph_updates.length}`);
  } catch (err: any) {
    console.log(`   -> Consolidation skipped or failed (ensure API handles session consolidation): ${err.message}`);
  }

  console.log("\n=== Simulation Complete ===");
}

main().catch(console.error);

File: examples/agent_loop.py
import asyncio
from datetime import datetime
from remem.client import Memory
from remem.models import MemoryType

async def main():
    print("🧠 Starting remem Agent Loop Simulation...")
    
    # Initialize the client. We assume the REST API is running locally on port 7474
    # Make sure your REMEM_API_KEY environment variable matches what the server expects.
    async with Memory(
        project="test-agent-01",
        base_url="http://127.0.0.1:7474",
        timeout=10.0
    ) as memory:
        
        print("\n--- Day 1: Storing Observations ---")
        observations = [
            ("The user prefers dark mode in all applications.", MemoryType.PREFERENCE, ["ui", "preferences"]),
            ("The codebase is written in Rust and uses Axum for the API.", MemoryType.FACT, ["architecture", "backend"]),
            ("When deploying to production, always use the GHCR docker image.", MemoryType.PROCEDURE, ["deployment", "devops"]),
            ("The user loves coffee but hates tea.", MemoryType.FACT, ["preferences", "food"]),
        ]
        
        for content, mem_type, tags in observations:
            print(f"Storing: '{content}'")
            try:
                result = await memory.store(
                    content=content,
                    memory_type=mem_type,
                    tags=tags
                )
                print(f"  ✅ Stored with ID {result.id} | Importance: {result.importance}")
            except Exception as e:
                print(f"  ❌ Failed to store: {e}")
                
        print("\n--- Day 2: Recalling Context ---")
        query = "How should I deploy the backend, and what is it written in?"
        print(f"Agent wants to know: '{query}'")
        
        try:
            results = await memory.recall(query, limit=2)
            print(f"\nRecall returned {len(results)} results:")
            for i, res in enumerate(results):
                print(f"  {i+1}. [{res.memory_type.value}] {res.content}")
                print(f"     Tags: {res.tags} | Relevance/Similarity: {res.similarity:.2f}")
                if res.reasoning:
                    print(f"     Reasoning: {res.reasoning}")
        except Exception as e:
            print(f"❌ Failed to recall: {e}")
            
        print("\n--- Day 3: Consolidating Session ---")
        session_id = "session_x99"
        print(f"Triggering consolidation for session {session_id}...")
        try:
            report = await memory.consolidate(session_id)
            print(f"  ✅ Consolidation Report:")
            print(f"     New Facts: {report.new_facts}")
            print(f"     Updated Facts: {report.updated_facts}")
            if report.contradictions:
                print(f"     Contradictions found: {len(report.contradictions)}")
        except Exception as e:
            # Note: Consolidate might fail if the server's consolidation logic is not fully hooked up in v0.1
            print(f"  ℹ️ Consolidation status: {e}")

if __name__ == "__main__":
    asyncio.run(main())
